### PR TITLE
Implement Variable.from_numpy

### PR DESCRIPTION
**Pull request by colesbury**
```
Implements from_numpy using ATen tensors. Variable.from_numpy is a
convenient placeholder for the variant that returns Variables until we
merge Tensor and Variable.

The behavior is slightly changed:

 - from_numpy() on an empty array now returns an empty tensor instead of
   throwing an exception. The shape may not be preserved.
 - CharTensor(ndarray) used to throw an exception. It now copies the
   ndarray. Copying is implemented via ATen toType.
```

### CODE REVIEW

File: torch/csrc/utils/tensor_numpy.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+#define NO_IMPORT_ARRAY
+#define PY_ARRAY_UNIQUE_SYMBOL __numpy_array_api
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+#include <numpy/arrayobject.h>
+
+using namespace at;
+
+namespace torch { namespace utils {
+
+static std::vector<npy_intp> cast_numpy(IntList x) {
```
Comment (killeent) [inline torch/csrc/utils/tensor_numpy.cpp]: nit: can you make this indicate that we are casting for size/stride metadata?

File: torch/autograd/variable.py [hunk 2]
```diff
@@ -358,6 +359,21 @@ def __dir__(self):
         keys = variable_methods + attrs
         return sorted(keys)
 
+    # Numpy array interface, to support `numpy.asarray(tensor) -> ndarray`
+    def __array__(self, dtype=None):
+        if dtype is None:
+            return self.cpu().numpy()
```
Comment (killeent) [inline torch/autograd/variable.py]: If the variable is on the GPU, is the implicit CPU conversion here to be expected by the user?

**Review by killeent — COMMENTED** [review]
Mostly LGTM.

File: torch/autograd/variable.py (hunk 2 above)
Comment (colesbury) [reply torch/autograd/variable.py]: It's copied as-is from Tensor.py, but I also think this is a reasonable choice.

**Review by killeent — APPROVED** [review]
Cool. Nice work on getting more and more stuff converging!

**Pull request merged**